### DISCUSSION TITLE
A bug about vm module

### DISCUSSION THREAD

**Discussion by xiaotiananan** [nodejs/help#4698] [Q&A]
### Version

18.4.1

### Subsystem

vm

### What steps will reproduce the bug?

```js
const vm = require('vm');
// Constructing context
var contextobj = { globalVar:6 };
// Contextifying object
vm.createContext(contextobj);
// Calling for loop
const script = new vm.Script(`console.log(this.hasOwnProperty("hasOwnProperty"))`);
// Calling runInContext method
script.runInContext(contextobj);
```

"The console should output false, but the result is true"

### How often does it reproduce? Is there a required condition?

always

### What is the expected behavior?

The console should output false, but the result is true

### What do you see instead?

The console output true

**Comment by bnoordhuis** [comment]
> The console should output false, but the result is true

Can you elaborate _why_ you expect that?

**Comment by xiaotiananan** [comment]
> > The console should output false, but the result is true
> 
> Can you elaborate _why_ you expect that?

Because the output is false in the browser environment and false in the v8 environment, but true in the vm environment

**Comment by koshic** [comment]
@xiaotiananan if you want to see false, just call `script.runInThisContext();` - you are using the wrong context and getting the wrong results

**Comment by bnoordhuis** [comment]
@koshic has it right. Closing as not a bug, can convert to a discussion if needed (edit: done)

**Comment by xiaotiananan** [comment]
> @xiaotiananan if you want to see false, just call `script.runInThisContext();` - you are using the wrong context and getting the wrong results

If I want to create a new context to execute the script, what should I do. My purpose is not to execute the script in the current context. Why can't we get the same results in the new context as in the browser environment.

**Reply by koshic** [reply]
> > @xiaotiananan if you want to see false, just call `script.runInThisContext();` - you are using the wrong context and getting the wrong results
> 
> If I want to create a new context to execute the script, what should I do. My purpose is not to execute the script in the current context. Why can't we get the same results in the new context as in the browser environment.


These are 2 different questions: how can you solve your problem and why VM module doesn't have methods to create context objects very close to node / browser global context.

I can't answer #2, so if you want - you can describe your setup & requirements a bit more and we can try to find alternative way (child process / fork as an example).
